_studio/mfx_lib/scheduler/include/mfx_scheduler_core_thread.h

```diff
@@ -31,25 +34,56 @@ class mfxSchedulerCore;
 
 struct MFX_SCHEDULER_THREAD_CONTEXT
 {
-    MFX_SCHEDULER_THREAD_CONTEXT()
+    typedef std::function<void(MFX_SCHEDULER_THREAD_CONTEXT*)> SchedulerThreadFunc;
+
+    MFX_SCHEDULER_THREAD_CONTEXT() = delete;
+
+    MFX_SCHEDULER_THREAD_CONTEXT(mfxU32 id)
       : state(State::Waiting)
-      , pSchedulerCore(NULL)
-      , threadNum(0)
-      , threadHandle()
+      , threadNum(id)
+      , threadHandle([this]() {
+          start.get_future().wait();
+          if (threadFunc) threadFunc(this);
+      })
       , workTime(0)
       , sleepTime(0)
-    {}
+    {
+    }
+
+    void Start(SchedulerThreadFunc func)
+    {
+        threadFunc = func;
+        start.set_value();
+    }
+
+    void Stop(std::lock_guard<std::mutex>& /*lock*/)
+    {
+        state = MFX_SCHEDULER_THREAD_CONTEXT::Stopping;
+        taskAdded.notify_one();
+    }
+
+    ~MFX_SCHEDULER_THREAD_CONTEXT()
+    {
+        if (threadHandle.joinable())
+        {
+            if (!threadFunc) // meaning Start() was not called
+                start.set_value();
+            threadHandle.join();
+        }
+    }
 
     enum State {
-        Waiting, // thread is waiting for incoming tasks
-        Running  // thread is executing a task
+        Waiting,  // thread is waiting for incoming tasks
+        Running,  // thread is executing a task
+        Stopping, // thread is going to stop
     };
 
     State state;                       // thread state, waiting or running
-    mfxSchedulerCore *pSchedulerCore;  // pointer to the owning core
     mfxU32 threadNum;                  // thread number assigned by the core
-    std::thread threadHandle;          // thread handle
     std::condition_variable taskAdded; // cond. variable to signal new tasks
+    std::promise<void> start;
+    SchedulerThreadFunc threadFunc;
+    std::thread threadHandle;          // thread handle
 
     mfxU64 workTime;                   // integral working time
```

Comment:
shouldn't it be switched to std::chrono in future?

Reply:
probably yes. But that's definitely out of scope of this PR